src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -0,0 +1,246 @@
+            return Ok(dev.lock().await.get_device_info().await);
+        }
+        Err(anyhow!(
+            "device with specified ID hasn't been created. {}",
+            device_id
+        ))
+    }
+
+    async fn find_device(&self, device_config: &DeviceConfig) -> Option<String> {
+        for (device_id, dev) in &self.devices {
+            match dev.lock().await.get_device_info().await {
+                DeviceConfig::Block(config) => match device_config {
+                    DeviceConfig::Block(ref config_new) => {
+                        if config_new.path_on_host == config.path_on_host {
+                            return Some(device_id.to_string());
+                        }
+                    }
+                    _ => {
+                        continue;
+                    }
+                },
+                _ => {
+                    // TODO: support find other device type
+                    continue;
+                }
+            }
+        }
+        None
+    }
+
+    async fn create_device(&mut self, device_config: &DeviceConfig) -> Result<String> {
+        // device ID must be generated by manager instead of device itself
+        // in case of ID collision
+        let device_id = self.new_device_id()?;
+        let dev: ArcMutexBoxDevice = match device_config {
+            DeviceConfig::Block(config) => self
+                .create_block_device(config, device_id.clone())
+                .await
+                .context("failed to create device")?,
+            _ => {
+                return Err(anyhow!("invliad device type"));
+            }
+        };
+        // register device to devices
+        self.devices.insert(device_id.clone(), dev.clone());
+        Ok(device_id)
+    }
+
+    async fn create_block_device(
+        &mut self,
+        config: &BlockConfig,
+        device_id: String,
+    ) -> Result<ArcMutexBoxDevice> {
+        let mut block_config = config.clone();
+        block_config.id = device_id.clone();
+        // get hypervisor block driver
+        let block_driver = match self
+            .hypervisor
+            .hypervisor_config()
+            .await
+            .blockdev_info
+            .block_device_driver
+            .as_str()
+        {
+            // convert the block driver to kata type
+            VIRTIO_BLOCK_MMIO => KATA_MMIO_BLK_DEV_TYPE.to_string(),
+            VIRTIO_BLOCK_PCI => KATA_BLK_DEV_TYPE.to_string(),
+            _ => "".to_string(),
+        };
+        block_config.driver_option = block_driver;
+        // generate virt path
+        let current_index = self.shared_info.declare_device_index()?;
+        block_config.index = current_index;
+        let drive_name = get_virt_drive_name(current_index as i32)?;
+        block_config.virt_path = format!("/dev/{}", drive_name);
+        // get device host path
+        block_config.path_on_host = get_host_path("b".to_owned(), config.major, config.minor)
```

Comment:
what about a rawfile based block device, will this path_on_host  deal with this case?

Reply:
add a if statement here, if it is rawfile based block device, the path on host is not empty. Otherwise, we use get_host_path to handle that